CEP XXXX: A backwards-compatible repodata update strategy for repodata `v3` (#146)
conda/ceps: cep-XXXX.md

## Motivation

`repodata.json` files are central to the conda ecosystem. They are the main source of packaging metadata and inform solvers about the catalog of available packages and their dependency constraints. As such, innovation work often refrains from modifying it, and the format itself has seen very few changes over its lifetime. However, a few ongoing efforts will inevitably result in `repodata.json` modifications (conditional dependencies, optional dependency groups, non-conda dependencies, etc).

The main problem is the introduction of backwards incompatible changes. The obvious solution is to bump the `repodata_version` field (like it was done with [CEP 15](./cep-0015.md)). However, this is not desirable for existing channels, since it immediately prevents non-compatible clients from interacting with the channel. Since most clients would update via a new version available in the channel, it creates a chichen-and-egg problem that would significantly delay the introduction of new features and hinder adoption.

> There must be a strategy to introduce backwards incompatible changes without breaking existing channels. This CEP centralizes the discussion for the update strategy and consolidates that feedback into a concrete proposal.

Author: I think the Rationale and Specification sections can be changed in order here, and then it's clearer. I'll use this summary as the beginning of the specification though.

## Rationale

### Version metadata in `info`

Adding a new field in the `info` dictionary is backwards compatible, and can be used by clients to parse the necessary keys directly without having to traverse the whole document. The migration timestamp is useful for client messaging, like "the client is not recent enough to see all records in this channel, please update to ensure you can obtain access to all packages". They are not added as a top-level field to stop polluting the global namespace.

### Using top-level fields for new metadata schemas

Adding new fields is backwards compatible and does not break older clients, which will simply ignore those and continue operating as usual.

### Freezing `repodata_version` to `1`

Bumps in this number should only result in backwards incompatible changes that would anyway prevent a channel from operating completely. While `repodata_version: 2` exists (as per CEP 15), its implementations are not sufficiently old to guarantee that the majority of existing conda clients would support it:

> - `rattler` supports it [v0.9.0](https://github.com/conda/rattler/blob/main/CHANGELOG.md#090---2023-09-22) (released on 2023-09-22), which means that `pixi` supports it since [v0.4.0](https://github.com/prefix-dev/pixi/blob/d8d2d8a3e8e1ce99707885aa1437e3768614456b/Cargo.toml#L38) (released on 2023-09-22 too).

Reviewer: missing "since"

```js
{
  "repodata_version": 1,
  "info": {
    "subdir": "noarch",
    "repodata_revisions": [
      {
        "revision": 3,
        "migrated_at": 1768249989751,  // 2026-01-12 20:33 UTC
      }
    ]
  },
  "packages": {
    "example-1.0.0-0.tar.bz2": {
      "build": "0",
      "build_number": 0,
      "depends": [],
      "md5": "82ecc40f09b9c44483e6b70cad2545d7",
      "name": "example",
      "noarch": "generic",
      "sha256": "eb65e866067865793b981c2ba74485f75bef441842b5998badc4ec66717685c7",
      "size": 1234,
      "subdir": "noarch",
      "timestamp": 1689209309623,
      "version": "1.0.0"
    }
  },
  "packages.conda": {
    "example-1.0.0-0.conda": {
      "build": "0",
      "build_number": 0,
      "depends": [],
      "md5": "82ecc40f09b9c44483e6b70cad2545d7",
      "name": "example",
      "noarch": "generic",
      "sha256": "eb65e866067865793b981c2ba74485f75bef441842b5998badc4ec66717685c7",
      "size": 1234,
      "subdir": "noarch",
      "timestamp": 1689209309623,
      "version": "1.0.0"
    }
  },
  "v3": {
    "tar.bz2": {},

>     "conda": {

Reviewer 2: I don't think this matters, and shards are not going to go away. For parsers, it will be easier to uniformly deal with the different version keys, then a new format, I feel.

>     "conda": {

Reviewer 3: I've gotten used to the v3 -> format -> package without extension and I kindof like it.

The main problem is the introduction of backwards incompatible changes. The obvious solution is to bump the `repodata_version` field (like it was done with [CEP 15](./cep-0015.md)). However, this is not desirable for existing channels, since it immediately prevents non-compatible clients from interacting with the channel. Since most clients would update via a new version available in the channel, it creates a chicken-and-egg problem that would significantly delay the introduction of new features and hinder adoption.

There must be a strategy to introduce backwards incompatible changes without breaking existing channels. This CEP centralizes the discussion for the update strategy and consolidates that feedback into a concrete proposal.

## Specification

A repodata revision introduces backwards incompatible features in a way that does not disrupt existing metadata. To do so, a new CEP MUST be proposed, following these guidelines below.

> A new item MUST be added to the `info.repodata_revisions` array, that MUST list the revisions found in the repodata file as dictionaries with the following schema:

Reviewer 4: Is there a reason to make this an array rather than a dict-of-dicts, with `vN` as the key and the other fields specified below as sub-dicts; e.g.,
```
"info": {
   ...
   "repodata_revisions": {
      "v3": {
        "n_packages": 42,
        "oldest": ....,
        "newest": ....
      },
      "v4": {
        "n_packages": 42,
        "oldest": ....,
        "newest": ....
      },
   }
   ...
}
```

A big issue I can see with `repodata_revisions` an array is that it (by itself) imposes no uniqueness constraint on `revision`, which leads us to have to answer, "what should tools do if they see multiple `{"revision": 3, ...}` in the array?"

Author: See [e65a5c5](https://github.com/conda/ceps/pull/146/commits/e65a5c5903829332c08e71f106420a85bb8640fa)

* `revision: int`: Required. The integer identifying the revision.

> * `n_packages: int`: Required. The number of packages available in this revision.

Reviewer 4: Since this is an required element, we should specify if tools MUST do anything particular with this data (e.g., verify the number of packages in the `vN` field), or is this field only exists for tools to use as they see fit (e.g., making package count verification optional).

Reviewer 2: maybe we should make it optional? I think it's purely informational

Author: See [e65a5c5](https://github.com/conda/ceps/pull/146/commits/e65a5c5903829332c08e71f106420a85bb8640fa)

* `oldest: int | None`: Optional. The timestamp (in milliseconds) of the oldest record published in this revision. If set to `None` or missing, the timestamp information is not available.

> * `newest: int | None`: Optional. The timestamp (in milliseconds) of the newest record published in this revision. If set to `None` or missing, the timestamp information is not available.

Reviewer 4: We could be a little clearer and say: "The [Unix time](https://en.wikipedia.org/wiki/Unix_time) in milliseconds" (a la CEP 34).

Reviewer 3: I'm unsure about tracking these numbers and the counts and propagating them back from individual shards generation to the shards index which has the info section. If we imagine a shards-only repodata, then the client would have to download all shards to double check package counts and timestamps.
Allowing conda-index to say "just pull all newer timestamps than X to the v3 section" would be the sort of implementation that I would be interested in, where conda-index would stay simple.
As a client, a boolean "client should be upgraded because there's repodata we don't understand?" is the most interesting information.

Reviewer 5: I agree with @dholth here.

Author: I made these values local to the file only. See [this comment](https://github.com/conda/ceps/pull/146#issuecomment-4383521254) too.

> * `newest: int | None`: Optional. The timestamp (in milliseconds) of the newest record published in this revision. If set to `None` or missing, the timestamp information is not available.

Reviewer 4: Also, not sure if we need to explicitly say these here, but:

* I assume timestamp ranges for various repodata revisions are allowed to overlap? E.g., the existing (v1/v2) `packages.conda` section can continue to publish packages, independent of any v3 date range.
* If newest and oldest are both present and not None, newest must be greater than or equal to oldest.

Reviewer 2: makes sense

Author: I think that's implied by the commonly understood meaning of "oldest" and "newest". Still, I reworded this part to ensure that it matches actual metadata in the file.

A new top-level field identified by the syntax `v{revision}` (where `revision` comes from `info.repodata_revisions[*].revision`) MUST map to a dictionary whose schema is presented in the relevant CEP.

> The CEP MUST also specify how to identify whether a given record belongs in the newer version, or can be added to the previous ones (e.g. a new field extending CEP 20's `info/index.json`)

Reviewer 4: Does "The CEP" here mean _this_ CEP, or later "implementation" CEPs making repodata revisions under this CEP?

Author: This is hopefully clearer now, PTAL 🙏

  - `newest: int | None`: If present and not `None`, a timestamp (in milliseconds) that MUST match the `indexed_timestamp` field of the newest record published in this revision in the current repodata file or shard.
  - `oldest: int | None`: If present and not `None`, a timestamp (in milliseconds) that MUST match the `indexed_timestamp` field of the oldest record published in this revision in the current repodata file or shard.

The `info.repodata_version` value MUST be `1` or, if [CEP 15](./cep-0015.md) applies, `2`.

### The `v3` top-level key

This key MUST map to a dictionary of type `dict[str, dict]`:

- Each key MUST be a non-empty string. The key SHOULD represent the file extension (without the leading period) of the included artifacts (usually `tar.bz2` and `conda`).
- Each value MUST be a dictionary of type `dict[str, dict]` where:
  - Each subkey MUST be a non-empty string representing the artifact filename without its extension.
  - Each subvalue MUST be a valid [CEP 36](./cep-0036.md) "package record metadata" dictionary,  including these changes:

>     - The `indexed_timestamp` field introduced by CEP PR [#154](https://github.com/conda/ceps/pull/154) SHOULD be set.

Author: Question for the audience (and particularly @jezdez): `SHOULD` or `MUST` here?

Reviewer 6: CEP 47 says that the field is optional and does not rely on this CEP for backwards compatibility. I think for consistency we should keep it as `SHOULD`.

    - The `extra_depends` field introduced by [44](./cep-0045.md) MAY be present.

>     - The MatchSpec strings mentioned in the fields `depends`, `constrains` and the lists of strings within `extra_depends` groups:

Reviewer 6: ```suggestion
    - The `extra-depends` field introduced by [44](./cep-0045.md) MAY be present.
    - The MatchSpec strings mentioned in the fields `depends`, `constrains` and the lists of strings within `extra-depends` groups:
```

Author: Yea, we need to get back to this. I think we need to amend that in CEP 45. Other fields in the record (e.g. `build_number`, `track_features`) use underscores as separators, so `extra-depends` is a clear outlier with that dash.

Reviewer 6: ya, I agree that amending CEP 45 would work as well. We are actually generating repodata on the conda-pypi channel with `extra_depends`, so changing it in the CEP would align the standard with what is being used in practice as well.

      - MUST set the `name` field to an exact string (no globbing allowed).

>       - MAY set the fields: `version`, `build`, `build_number`, CEP 43's `when`, CEP 44 `extra`, CEP 45 `flags`.

Reviewer 6: ```suggestion
      - MAY set the fields: `version`, `build`, `build_number`, CEP 43 `when`, CEP 44 `extras`, CEP 45 `flags`.
```

      - MUST NOT set any other fields.

>       - MUST be represented with the `name` + square-brackets form (e.g. `name[version="1.2.*",build_number=0]`)

Author: Reviewers, please pay attention to this part, since it restricts which `MatchSpec` strings can be used.

Reviewer 6: Right now with v3 repodata we are generating with conda-index, version doesn't appear in square brackets. I understand that this CEP as worded would require `pytest[version=">=2.7.2"]` instead of `pytest>=2.7.2`. Is that a correct understanding and can you help me understand the motivation to make the syntax more restrictive? Thanks!

Reviewer 5: Also curious about this!

Author: Non-square brackets syntax can be very ambiguous (especially when build strings are also added). I want to move away from that syntax in stored repodata (and leave it as syntactic sugar for user input only). Users can still add whatever in the recipes (that bit is not specified here), but repodata creation tools (indexing) must canonicalize the representation in the repodata for clarity.

Is that a good enough reason?

Author: cc @baszalmstra @wolfv

>       - MUST be represented with the `name` + square-brackets form (e.g. `name[version="1.2.*",build_number=0]`)

Reviewer 6: ```suggestion
      - If only `name` is set, the MatchSpec MUST be the bare name string (e.g. `pip`). Empty brackets (e.g. `pip[]`) MUST NOT be used.
      - If any field other than `name` is set, the MatchSpec MUST use the `name` + square-brackets form (e.g. `name[version="1.2.*",build_number=0]`).
```

Does this help clarify the intent instead of a catch-all bracket bullet?

Reviewer 2: Makes sense to me.

Author: Applied, thanks.

        "timestamp": 1768249989851,
        "version": "2.0.0",
      },
      "example-3.0.0-0": {  // key does not have the extension anymore
        "build": "0",
        "build_number": 0,
        "depends": [
          "package[version=3,build_number=0,when=__unix]"  // bracket syntax, w/ conditional
        ],
        "md5": "6b70cad2545d782ecc40f09b9c44483e",
        "name": "example",
        "noarch": "generic",
        "sha256": "74485f75bef441842b5998badc4ec66717685c7eb65e866067865793b981c2ba",
        "size": 2345,
        "subdir": "noarch",
        "timestamp": 1773851561010,
        "version": "3.0.0",
      }
    }
  }
}
```

## Rejected ideas

### Encoding version information in the filename

One alternative would be to create new `repodata.json` filenames (e.g. `repodata.v4.json`) for each new incompatible bump. However, this was rejected by the authors as it introduces complexity in other areas:

- It would require more HTTP calls to retrieve the latest version served by the channel.
- It would introduce duplication across `repodata.json` versions and their shards.
- Indexing tools would need to maintain the different versions of `repodata.json` in sync.

> ### Nested `name`-`version`-`build_string` dictionaries

Author: I honestly think this is a good idea and so far I don't have any solid arguments to say no. Too much nesting? Taste?

Reviewer 5: It makes sparsely parsing the repodata a bit less efficient due to the nested structure. And what is the benefit? This file is not for human consumption anyway.

Author: @dholth hinted that it would be better suited to find packages by name, instead of parsing strings from the flattened level. But no measurements were provided. If it hinders sparse-parsing, then I guess that's a good argument to include.

Reviewer 3: I am currently anti-nesting.

A review comment proposed subdividing the dictionary of filenames to records in nested dictionaries. In other words, instead of listing the artifact `package-version-build.conda` as:

```js
{
  ...,
  "v3": {
    "conda": {
      "example-2.0.0-0": {
        ... // record dictionary
      }
    }
  }
}
```

... expose it as:

```js
{
  ...,
  "v3": {
    "conda": {      // extension
      "example": {  // name
        "2.0.0":    // version
          "0": {    // build string
            ...     // record dictionary
          }
        }
      }
    }
  }
}
```

This would simplify parsing of records, but would also complicate the enumeration of existing records.

> ### Expose `MatchSpec` entries in records as dictionaries 

Author: I also like this one by @baszalmstra a lot; why aren't we doing this?

Reviewer 5: We are designing this for pixi as well, but with toml. We are struggling to find a good way to define the recursive boolean expression logic that can exist in when statements nicely.

Author: Yea I guess that `when` would remain as a string in this case.

Bumps in this number should only result in backwards incompatible changes that would anyway prevent a channel from operating completely. While `repodata_version: 2` exists (as per CEP 15), its implementations are not sufficiently old to guarantee that the majority of existing conda clients would support it:

- `rattler` supports it since [v0.9.0](https://github.com/conda/rattler/blob/main/CHANGELOG.md#090---2023-09-22) (released on 2023-09-22), which means that `pixi` supports it since [v0.4.0](https://github.com/prefix-dev/pixi/blob/d8d2d8a3e8e1ce99707885aa1437e3768614456b/Cargo.toml#L38) (released on 2023-09-22 too).
- `conda` only supports it as of [v24.5.0](https://github.com/conda/conda/blob/main/CHANGELOG.md#2450-2024-05-08) (released on 2024-05-08)
- `mamba` started supporting it in [v2.0](https://github.com/mamba-org/mamba/blob/main/CHANGELOG.md#20240925) (released on 2024-09-25).

Hence, we suggest to stick to `repodata_version: 1` and _only_ use `repodata_version: 2` when a new channel needs a global `base_url` for all the entries in the `packages` and `packages.conda` fields.

## Example

```js
{
  "repodata_version": 1,
  "info": {
    "subdir": "noarch",

>     "repodata_revisions": {

Reviewer 6: I noticed that py-rattler 0.24.0 now implements this as a list of objects this an integer revision field like:
`"repodata_revisions": [{"revision": 3, "n_packages": 2, "oldest": ..., "newest": ...}]`

Reviewer 5: Yeah I think this was not updated yet to follow the latest changes in the CEP. Are you currently relying on this? Happy to implement this ASAP.

Reviewer 5: The previous version of py-rattler didnt parse this at all.

Reviewer 2: I think this can be resolved now :)
